proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,489 @@
+
+        self.handler.log(level: level,
+                         message: message(),
+                         metadata: metadata,
+                         source: source() ?? Logger.currentModule(fileID: (file)),
+                         file: file, function: function, line: line)
+    }
+}
+```
+
+Note, that the task-local lookup is only performed when the log message is about to be emitted. No task-local or thread-local lookups are performed when the log level is lower than the level of the current statement or in case no metadata provider was set up.
+
+We also introduce another overload of the log (and trace/debug/info/notice/warning/error/critical) method, which allows libraries to pass a baggage explicitly if they need to (or, disable baggage value extraction by passing `nil` as the value):
+
+```swift
+extension Logger {
+    @inlinable
+    public func log(level: Logger.Level,
+                    _ message: @autoclosure () -> Logger.Message,
+                    metadata: @autoclosure () -> Logger.Metadata? = nil,
+                    baggage: Baggage?,
+                    source: @autoclosure () -> String? = nil,
+                    file: String = #fileID, function: String = #function, line: UInt = #line) {
+        if self.logLevel <= level {
+            let providerMetadata = metadataProvider?.metadata(baggage)
+            let metadata: Metadata? = {
+                guard let metadata = metadata(), !metadata.isEmpty else { return providerMetadata }
+                guard let providerMetadata = providerMetadata else { return metadata }
+                return providerMetadata.merging(metadata, uniquingKeysWith: { _, oneOffMetadata in oneOffMetadata })
+            }()
+
+            self.handler.log(level: level,
+                             message: message(),
+                             metadata: metadata,
+                             source: source() ?? Logger.currentModule(fileID: (file)),
+                             file: file, function: function, line: line)
+        }
+    }
+}
+```
+
+Practical usage of baggage values is as follows:
+
+```swift
+var baggage = Baggage.topLevel
+baggage[TestIDKey.self] = "my-awesome-trace-id"
+
+Baggage.$current.withValue(baggage) {
+    logger.info("Such logging, wow!", metadata: ["one-off": "example"])
+    // info [my-test-key: my-awesome-trace-id, one-off: example] Such logging, wow!
+}
+```
+
+This has the advantage of not having to specify the global metadata explicitly anymore.
+
+The task-local baggage is often going to be set by e.g. tracing libraries or frameworks, so end-users do not need to do anything to benefit from tracing,
+other than setting up the metadata providers when they bootstrap their logging and tracing systems. For example, assuming some imaginary http framework where http requests are forwarded to an actor for handling, all the end-users of this framework would benefit from contextual metadata, propagated with baggage by logging, as usual:
+
```

Comment:
Just for my full understanding. Let's take the example of a gRPC-server. What we expect the gRPC library folks to implement is once they decode the trace-id from the incoming requests headers to set them on the current task local baggage right?
What key would they use to set on the `Baggage`? Is that key coming from `swift-distributed-tracing`?

Reply:
I'll add more examples for the server side, tho tbh that perhaps is time better spent on revamping the tracing "what you should do" guide and I'll link to it from here rather in this proposal itself. I'll add more examples tho, good question - thanks.

Baggage has _ONE_ "blessed" task-local key and it is the `Baggage.current`; that's what an instrumented system would be setting.